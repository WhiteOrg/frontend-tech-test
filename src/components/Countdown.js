import React, { useState, useEffect, useCallback } from "react";
import { Rectangle, CountdownContainer } from "../styles/header";
import { countdownDate, msInADay } from "../constants/constants";

const Countdown = () => {
  const [timeleft, setTimeLeft] = useState({});
  const { days, hours, minutes, seconds } = timeleft;

  const calculateTimeLeft = useCallback(() => {
    let diff = countdownDate - new Date().getTime();
    if (diff > 0) {
      const timeLeftFormatted = {
        days: Math.floor(diff / msInADay),
        hours: Math.floor((diff / (1000 * 60 * 60)) % 24),
        minutes: Math.floor((diff / 1000 / 60) % 60),
        seconds: Math.floor((diff / 1000) % 60),
      };
      return timeLeftFormatted;
    } else {
      return {
        days: 0,
        hours: 0,
        minutes: 0,
        seconds: 0,
      };
    }
  }, []);

  useEffect(() => {
    const interval = setInterval(() => {
      setTimeLeft(calculateTimeLeft());
    }, 1000);
    return () => clearInterval(interval);
  }, [calculateTimeLeft]);

  return (
    <CountdownContainer>
      <p>Starts</p>
      <Rectangle>{days}</Rectangle>
      <span>d</span>
      <Rectangle>{hours}</Rectangle>
      <span>h</span>
      <Rectangle>{minutes}</Rectangle>
      <span>m</span>
      <Rectangle>{seconds}</Rectangle>
      <span>s</span>
    </CountdownContainer>
  );
};

export default Countdown;
