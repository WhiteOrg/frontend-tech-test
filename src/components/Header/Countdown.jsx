import React, { useState } from 'react'
import styled from 'styled-components'



const Counter = styled.div`
    width: 100%;
    display: flex;
    justify-content: space-around;
    align-items: baseline;
`

const CounterTypo = styled.h3`
    color: #fff;
    min-width: 80px;
    font-weight: 500;
    font-size: 1.5em;
    margin-right: 0.3em;

`
const CounterItems = styled.div`
    display: flex;
    flex: 1;
`

const CounterDigit = styled.p`
    width: 1.8em;
    height: 1.8em;
    display: flex;
    justify-content: center;
    align-items: center;
    color: #fff;
    font-size: 1.2em;
    font-weight: 500;
    border: 2px solid #fff;
    border-radius: 8px;
    position: relative;
    z-index: 3;

    ::before {
        content: "";
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%,-50%);
        width: 90%;
        height: 3px;
        background-color: rgba(255, 255, 255, 0.3);
        border: none;
        border-radius: 2px;
        z-index: 1;
    }
`

const CounterLetter = styled(CounterDigit)`
    border: none;
    border-radius: 0;
    ::before {
        display: none;
    }
`


// JSX
export default function Countdown() {

// SETTING UP STATE VARIABLES FOR THE TIMER
const [timerDays, setTimerDays] = useState('00')
const [timerHours, setTimerHours] = useState('00')
const [timerMinutes, setTimerMinutes] = useState('00')
const TIMER_CYCLE_TIME = 1000



let interval = React.useRef()

// SO FOR THIS EXAMPLE I DID 7 DAYS SINCE DATE.NOW()
// NORMALLY I WOULD FETCH END-DATE FROM API AND COMPARE TO CURRENT DATE
// LIKE SO
// 
// GetDataFromAPI().then( data => setEndDate( new Date(data.endDate) ))
// 
// UNFORTUNATELY THERE WAS AN INVALID DATE FORMAT FOR ENDDATE IN API OBJECT
// 2020-15-06T10:06:29.578Z -> INVALID DATE AS THERE IS NO 15TH MONTH IN THE YEAR
// 2020-06-15T10:06:29.578Z -> I THINK THIS WAS A GOAL? 


// CREATED FUNCTION INSIDE THE EFFECT HOOK AS I DONT USE IT ANYWHERE ELSE SO ITS FINE IN THIS SCENARIO
// AND ALSO AVOIDED SOME WARNIGS INSIDE CONSOLE ABOUT MISSING DEPENDANCIES

React.useEffect( () => {
    // SETUP START DAY 
    const startDate = new Date()
    startDate.setDate(startDate.getDate() + 7 )

    const startTime = () => {

        interval.current = setInterval(() => {
            const now = new Date().getTime()
            const distance = startDate - now
    
            const days = Math.floor( distance / ( 1000 * 60 * 60 * 24 ) )
            const hours = Math.floor( distance % ( 1000 * 60 * 60 * 24) / ( 1000 * 60 * 60 ) )
            const minutes = Math.floor( distance % ( 1000 * 60 * 60 ) / ( 1000 * 60 ) )
    
            // IF DISTANCE IS LESS THAN 0 -> TIME IS UP
            if (distance < 0) {
                // STOP THE TIMER
                clearInterval( interval.current)
            } else {
                // UPDATE THE TIMER
                setTimerDays( days )
                setTimerHours( hours )
                setTimerMinutes( minutes )
            }
        }, TIMER_CYCLE_TIME)
        
    }    // SET IT TO VARIABLE SO IT UPDATES WHEN INTERVAL HAS CHANGED
    let toClean = interval.current

    startTime()

    // CLEAN UP FUNCTION 
    return () => clearInterval( toClean )

},[])

    return(
        <Counter>
            <CounterTypo>Starts at</CounterTypo>
            <CounterItems>
                <CounterDigit>{ timerDays }</CounterDigit>
                <CounterLetter>d</CounterLetter>
                <CounterDigit>{ timerHours }</CounterDigit>
                <CounterLetter>h</CounterLetter>
                <CounterDigit>{ timerMinutes }</CounterDigit>
                <CounterLetter>m</CounterLetter>
            </CounterItems>
    </Counter>
    )
}