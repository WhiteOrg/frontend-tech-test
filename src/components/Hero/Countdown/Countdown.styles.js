import styled from "styled-components";

export const Container = styled.div`
  font-family: ${({ theme }) => theme.fonts.primary};
  color: #fff;
  font-size: 1.4rem;
  font-weight: 500;

  @media ${({ theme }) => theme.mediaQueries["tablet"]} {
    font-size: 1.6rem;
  }

  @media ${({ theme }) => theme.mediaQueries["laptop"]} {
    font-size: 1.8rem;
  }

  @media ${({ theme }) => theme.mediaQueries["desktop"]} {
    font-size: 2rem;
  } ;
`;

export const Elements = styled.span`
  display: inline-flex;
  margin: 0 4px;
`;

export const Span = styled.span`
  border-radius: 5px;
  display: inline-flex;
  justify-content: center;
  align-items: center;

  &:nth-child(odd) {
    border: 2px solid #fff;
    height: 35px;
    width: 30px;
    color: #fff;
    position: relative;
    z-index: 1;
    margin-right: 8px;

    @media ${({ theme }) => theme.mediaQueries["tablet"]} {
      height: 38px;
      width: 27px;
    }
    @media ${({ theme }) => theme.mediaQueries["laptop"]} {
      height: 40px;
      width: 35px;
    }
    @media ${({ theme }) => theme.mediaQueries["desktop"]} {
      height: 45px;
      width: 40px;
    }
    &::after {
      content: "";
      width: 80%;
      border-bottom: 4px solid #4e757d;
      border-radius: 1rem;
      position: absolute;
      left: 10%;
      top: 50%;
      z-index: -1;
    }
  }
`;
