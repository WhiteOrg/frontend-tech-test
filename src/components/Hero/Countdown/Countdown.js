import React, { useState, useEffect } from "react";
import { Container, Span, Elements } from "./Countdown.styles";

const targetDate = new Date();
targetDate.setDate(targetDate.getDate() + 7);

const Countdown = () => {
  const calculateTimeLeft = () => {
    let getCurrentTime = new Date();
    let difference = targetDate - getCurrentTime;

    let timeLeft = {};
    if (difference > 0) {
      timeLeft = {
        d: Math.floor(difference / (1000 * 60 * 60 * 24)),
        h: Math.floor((difference / (1000 * 60 * 60)) % 24),
        m: Math.floor((difference / 1000 / 60) % 60),
      };
    }
    return timeLeft;
  };
  const [timeLeft, setTimeLeft] = useState(calculateTimeLeft());

  useEffect(() => {
    const timer = setTimeout(() => {
      setTimeLeft(calculateTimeLeft());
    }, 60 * 1000);

    return () => clearTimeout(timer);
  });

  const timerComponents = [];
  Object.keys(timeLeft).forEach((interval, index) => {
    timerComponents.push(
      <Elements key={index}>
        <Span>{timeLeft[interval]}</Span> <Span>{interval} </Span>
      </Elements>
    );
  });

  return <Container>Starts at {timerComponents}</Container>;
};

export default Countdown;
