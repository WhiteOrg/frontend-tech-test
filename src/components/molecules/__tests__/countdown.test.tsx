import React from "react";
import renderer from "react-test-renderer";
import Countdown from "../countdown";

describe("MOLECULES: Render testing", () => {
  test("Should render Countdown", () => {
    let startDate = "2020-08-08T10:06:29.578Z";
    const component = renderer.create(
      <div>
        <Countdown startDate={startDate} />
      </div>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
