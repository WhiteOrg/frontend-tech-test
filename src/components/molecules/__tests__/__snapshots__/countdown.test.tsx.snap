// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ATOMS: Render testing Should render Countdown 1`] = `
<div>
  <div
    className="sc-AxhCb fzQUN"
  >
    <p
      className="sc-AxhUy blACGn"
    >
      Start at
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      d
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      h
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      m
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      s
    </p>
  </div>
</div>
`;

exports[`ATOMS: Render testing Should render Hr 1`] = `
<div>
  <div
    className="sc-AxhCb fzQUN"
  >
    <p
      className="sc-AxhUy blACGn"
    >
      Start at
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      d
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      h
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      m
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      s
    </p>
  </div>
</div>
`;

exports[`MOLECULES: Render testing Should render Countdown 1`] = `
<div>
  <div
    className="sc-AxhCb fzQUN"
  >
    <p
      className="sc-AxhUy blACGn"
    >
      Start at
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      d
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      h
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      m
    </p>
    <div
      className="sc-AxirZ hqRJJG"
    >
      <div
        className="sc-AxjAm crHSQi"
      />
      <p
        className="sc-AxiKw dPBvyy"
      >
        0
      </p>
    </div>
    <p
      className="sc-AxhUy blACGn"
    >
      s
    </p>
  </div>
</div>
`;
